### PR TITLE
move `all_zeroes` & `from_byte_array` to inherents & make them const.

### LINKED ISSUE #2377: const constructors for hashes
The hashes currently have no const constructors (not even `from_byte_array` or `all_zeros`).

I found out because I was trying to make `OutPoint::new` and `OutPoint::null` const fns, but the latter is impossible, it seems.

### PR DESCRIPTION
**Pull request by 15IITian**
Made these functions const fn & move them from the `Hash` trait to `hash_type` & `hash_newtype` macros

fix #2377

### CODE REVIEW

File: hashes/src/internal_macros.rs [hunk 1]
```diff
@@ -188,6 +180,18 @@ macro_rules! hash_type {
...
+
+            /// Constructs a hash from the underlying byte array.
+            pub const fn from_byte_array(bytes: <Self as crate::Hash>::Bytes) -> Self {
+                Hash(bytes)
+            }
+
+            /// Returns an all zero hash.
+            ///
+            /// An all zeros hash is a made up construct because there is not a known input that can create
+            /// it, however it is used in various places in Bitcoin e.g., the Bitcoin genesis block's
```
Comment (apoelstra) [inline hashes/src/internal_macros.rs:192]: Grammar nit: change `it, however` to `it. However, `. (comma to period, then add a comma after howenver).

I know the typo was in the original comment, but let's fix it now.

Comment (apoelstra) [reply hashes/src/internal_macros.rs:192]: You will need to change this in several other places as well.

**Comment by apoelstra** [comment]
1d6373df6a69b43603796ac2bb3794412dff1c5d looks good to me other than the grammar thing.

Also, the CI failure is minor but real -- please remove the now-unused `Hash` import from the example.

**Comment by 15IITian** [comment]
https://github.com/rust-bitcoin/rust-bitcoin/actions/runs/8490482071/job/23261584927?pr=2641

Can you please tell how to solve this error.

**Comment by 15IITian** [comment]
I noticed -> our code is not  formatted according to rust fmt tool (cargo fmt).

**Comment by tcharding** [comment]
Thanks for the PR, this is not a refactor, its an API breaking change because it removes functions from a public trait.  I always link contributors to https://chris.beams.io/posts/git-commit/ (also mentioned in CONTRIBUTING.md). If a dev describes _why_ we should have a change, and _what_ the change is then it clarifies a bunch of things, including that:
- dev understands fully the change (the why and the what)
- reviewer can understand the context

I don't think you would have written "refactor" after writing a why/what PR description for the changes in this PR :)

**Comment by tcharding** [comment]
Converting to draft because we are in a feature freeze right now, this needs to wait until `rust-bitcoin 0.32.0` is out.

**Comment by 15IITian** [comment]
> this is not a refactor, its an API breaking change because it removes functions from a public trait

Well ,I agree that it  is more often a api breaking change but  while writting conventional commit ->I could'nt find more relevant tag than 'refactor' -> `A code change that neither fixes a bug nor adds a feature ` :)


> I always link contributors to https://chris.beams.io/posts/git-commit/ (also mentioned in CONTRIBUTING.md)

:+1:

**Comment by tcharding** [comment]
"Refactor" is a well defined term:

> In computer programming and software design, code refactoring is the process of restructuring existing computer code—changing the factoring—without changing its external behavior.

ref: https://en.wikipedia.org/wiki/Code_refactoring.

We don't enforce any particular format of git brief messages, so there is no requirement to put any sort of tag at the start of the message. Just something simple like "Use inherent functions instead of trait methods" would be fine.

Thanks

**Comment by tcharding** [comment]
Feature freeze is over man, you are free to push this up if you want to.

**Comment by tcharding** [comment]
I know you did this ages ago, and sorry to waste your time, but I think this PR is not required now because of the `hashes` re-write in #2770. I note that I had forgotten the `const` stuff in that PR until reviewing this today (and have now update it) so your effort was not wasted.

**Comment by 15IITian** [comment]
No problem mate.
Good to see , that I have indirectly contributed in hashes pr.

**Pull request closed without merging**